### ISSUE TITLE
[Bug]: Error in forward pass to compute stress

### ISSUE THREAD

**Issue by ajhoffman1229** [materialsvirtuallab/matgl#171]
### Email (Optional)

ajhoff29@mit.edu

### Version

0.8.5

### Which OS(es) are you using?

- [ ] MacOS
- [ ] Windows
- [X] Linux

### What happened?

I'm attempting to run the code in the snippet below to compute the stress on a siliceous CHA unit cell with the ASE M3GNetCalculator (CIF file available [here](https://america.iza-structure.org/IZA-SC/cif/CHA.cif), but I think these issues arise with any ASE Atoms object). I did two things that seemed to resolve the issue:

- changed on line 61 in `matgl/layers/_three_body.py` to `weights = three_cutoff[torch.stack(list(line_graph.edges()), dim=1).long()].view(-1, 2)`
- changed line 64 in `matgl/layers/_atom_ref.py` to `one_hot = torch.eye(num_elements)[g.ndata["node_type"].long()]`.

Thanks in advance for your help! Looks like there were just a few spots where the tensors needed to be changed to `.long()`.

EDIT: I made different changes to the code than I had in my original bug report, apologies for the mistake.

### Code snippet

```python
import matgl
from matgl.ext.ase import M3GNetCalculator
potential = matgl.load_model("M3GNet-MP-2021.2.8-PES")
calculator = M3GNetCalculator(potential=potential)
atoms.calc = calculator
atoms.get_stress()
```

### Log output

```shell
# === First Error ===

---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[1], line 13
     11 calculator = M3GNetCalculator(potential=potential)
     12 atoms.calc = calculator
---> 13 atoms.get_stress()

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/ase/atoms.py:820, in Atoms.get_stress(self, voigt, apply_constraint, include_ideal_gas)
    817 if self._calc is None:
    818     raise RuntimeError('Atoms object has no calculator.')
--> 820 stress = self._calc.get_stress(self)
    821 shape = stress.shape
    823 if shape == (3, 3):
    824     # Convert to the Voigt form before possibly applying
    825     # constraints and adding the dynamic part of the stress
    826     # (the "ideal gas contribution").

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/ase/calculators/abc.py:26, in GetPropertiesMixin.get_stress(self, atoms)
     25 def get_stress(self, atoms=None):
---> 26     return self.get_property('stress', atoms)

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/ase/calculators/calculator.py:737, in Calculator.get_property(self, name, atoms, allow_calculation)
    735     if not allow_calculation:
    736         return None
--> 737     self.calculate(atoms, [name], system_changes)
    739 if name not in self.results:
    740     # For some reason the calculator was not able to do what we want,
    741     # and that is OK.
    742     raise PropertyNotImplementedError('{} not present in this '
    743                                       'calculation'.format(name))

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/matgl/ext/ase.py:178, in M3GNetCalculator.calculate(self, atoms, properties, system_changes)
    176     energies, forces, stresses, hessians = self.potential(graph, self.state_attr)
    177 else:
--> 178     energies, forces, stresses, hessians = self.potential(graph, state_attr_default)
    179 self.results.update(
    180     energy=energies.detach().numpy(),
    181     free_energy=energies.detach().numpy(),
    182     forces=forces.detach().numpy(),
    183 )
    184 if self.compute_stress:

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/torch/nn/modules/module.py:1194, in Module._call_impl(self, *input, **kwargs)
   1190 # If we don't have any hooks, we want to skip the rest of the logic in
   1191 # this function, and just call forward.
   1192 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1193         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1194     return forward_call(*input, **kwargs)
   1195 # Do not call functions when jit is used
   1196 full_backward_hooks, non_full_backward_hooks = [], []

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/matgl/apps/pes.py:76, in Potential.forward(self, g, state_attr, l_g)
     73 if self.calc_forces:
     74     g.ndata["pos"].requires_grad_(True)
---> 76 predictions = self.model(g, state_attr, l_g)
     77 if isinstance(predictions, tuple) and len(predictions) > 1:
     78     total_energies, site_wise = predictions

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/torch/nn/modules/module.py:1194, in Module._call_impl(self, *input, **kwargs)
   1190 # If we don't have any hooks, we want to skip the rest of the logic in
   1191 # this function, and just call forward.
   1192 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1193         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1194     return forward_call(*input, **kwargs)
   1195 # Do not call functions when jit is used
   1196 full_backward_hooks, non_full_backward_hooks = [], []

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/matgl/models/_m3gnet.py:252, in M3GNet.forward(self, g, state_attr, l_g)
    250 node_feat, edge_feat, state_feat = self.embedding(node_types, g.edata["rbf"], state_attr)
    251 for i in range(self.n_blocks):
--> 252     edge_feat = self.three_body_interactions[i](
    253         g,
    254         l_g,
    255         three_body_basis,
    256         three_body_cutoff,
    257         node_feat,
    258         edge_feat,
    259     )
    260     edge_feat, node_feat, state_feat = self.graph_layers[i](g, edge_feat, node_feat, state_feat)
    261 g.ndata["node_feat"] = node_feat

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/torch/nn/modules/module.py:1194, in Module._call_impl(self, *input, **kwargs)
   1190 # If we don't have any hooks, we want to skip the rest of the logic in
   1191 # this function, and just call forward.
   1192 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1193         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1194     return forward_call(*input, **kwargs)
   1195 # Do not call functions when jit is used
   1196 full_backward_hooks, non_full_backward_hooks = [], []

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/matgl/layers/_three_body.py:61, in ThreeBodyInteractions.forward(self, graph, line_graph, three_basis, three_cutoff, node_feat, edge_feat)
     59 print(three_cutoff)
     60 print(torch.stack(list(line_graph.edges()), dim=1).view(-1, 2))
---> 61 weights = three_cutoff[torch.stack(list(line_graph.edges()), dim=1)].view(-1, 2)  # type: ignore
     62 weights = torch.prod(weights, dim=-1)  # type: ignore
     63 basis = basis * weights[:, None]

IndexError: tensors used as indices must be long, byte or bool tensors

# === Second Error ===

---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[1], line 13
     11 calculator = M3GNetCalculator(potential=potential)
     12 atoms.calc = calculator
---> 13 atoms.get_stress()

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/ase/atoms.py:820, in Atoms.get_stress(self, voigt, apply_constraint, include_ideal_gas)
    817 if self._calc is None:
    818     raise RuntimeError('Atoms object has no calculator.')
--> 820 stress = self._calc.get_stress(self)
    821 shape = stress.shape
    823 if shape == (3, 3):
    824     # Convert to the Voigt form before possibly applying
    825     # constraints and adding the dynamic part of the stress
    826     # (the "ideal gas contribution").

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/ase/calculators/abc.py:26, in GetPropertiesMixin.get_stress(self, atoms)
     25 def get_stress(self, atoms=None):
---> 26     return self.get_property('stress', atoms)

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/ase/calculators/calculator.py:737, in Calculator.get_property(self, name, atoms, allow_calculation)
    735     if not allow_calculation:
    736         return None
--> 737     self.calculate(atoms, [name], system_changes)
    739 if name not in self.results:
    740     # For some reason the calculator was not able to do what we want,
    741     # and that is OK.
    742     raise PropertyNotImplementedError('{} not present in this '
    743                                       'calculation'.format(name))

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/matgl/ext/ase.py:178, in M3GNetCalculator.calculate(self, atoms, properties, system_changes)
    176     energies, forces, stresses, hessians = self.potential(graph, self.state_attr)
    177 else:
--> 178     energies, forces, stresses, hessians = self.potential(graph, state_attr_default)
    179 self.results.update(
    180     energy=energies.detach().numpy(),
    181     free_energy=energies.detach().numpy(),
    182     forces=forces.detach().numpy(),
    183 )
    184 if self.compute_stress:

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/torch/nn/modules/module.py:1194, in Module._call_impl(self, *input, **kwargs)
   1190 # If we don't have any hooks, we want to skip the rest of the logic in
   1191 # this function, and just call forward.
   1192 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1193         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1194     return forward_call(*input, **kwargs)
   1195 # Do not call functions when jit is used
   1196 full_backward_hooks, non_full_backward_hooks = [], []

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/matgl/apps/pes.py:84, in Potential.forward(self, g, state_attr, l_g)
     82 total_energies = self.data_std * total_energies + self.data_mean
     83 if self.element_refs is not None:
---> 84     property_offset = torch.squeeze(self.element_refs(g))
     85     total_energies += property_offset
     87 forces = torch.zeros(1)

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/torch/nn/modules/module.py:1194, in Module._call_impl(self, *input, **kwargs)
   1190 # If we don't have any hooks, we want to skip the rest of the logic in
   1191 # this function, and just call forward.
   1192 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1193         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1194     return forward_call(*input, **kwargs)
   1195 # Do not call functions when jit is used
   1196 full_backward_hooks, non_full_backward_hooks = [], []

File ~/mambaforge/envs/htvs/lib/python3.9/site-packages/matgl/layers/_atom_ref.py:64, in AtomRef.forward(self, g, state_attr)
     52 """Get the total property offset for a system.
     53
     54 Args:
   (...)
     59     offset_per_graph
     60 """
     61 num_elements = (
     62     self.property_offset.size(dim=1) if self.property_offset.ndim > 1 else self.property_offset.size(dim=0)
     63 )
---> 64 one_hot = torch.eye(num_elements)[g.ndata["node_type"]]
     65 if self.property_offset.ndim > 1:
     66     offset_batched_with_state = []

IndexError: tensors used as indices must be long, byte or bool tensors
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by ajhoffman1229** [comment]
It seems there are still some issues with stress calculations in matgl. Even after I made the above changes, I get the following unusual outputs when I run two commands that should provide the same result:
```
atoms.calc = calculator
atoms.get_stress() # returns a 1D array with 6 entries, correct
# Out: array([-4.01619339e+00, -4.01619482e+00, -3.35417414e+00, -2.44878402e-06,
       -7.62686909e-07, -4.99140640e-08])

calculator.get_property("stress", atoms) # returns a 3x3 array
# Out: array([[-4.0161934e+00, -1.4574907e-07, -7.3473507e-07],
       [ 4.5920942e-08, -4.0161948e+00, -2.3459613e-06],
       [-7.9063881e-07, -2.5516069e-06, -3.3541741e+00]], dtype=float32)
```

For comparison, here is an example with another calculator that just evaluates the dispersive forces:
```
from dftd4.ase import DFTD4
dispcalc = DFTD4(method='pbe')
atoms.calc = dispcalc
atoms.get_stress()
# Out: array([ 2.91348683e-03,  2.91348685e-03,  3.06167218e-03,  4.54646702e-13,
       -7.51897679e-13,  2.23088171e-11])

dispcalc.get_property("stress", atoms) # correctly returns the same output as above
# Out: array([ 2.91348683e-03,  2.91348685e-03,  3.06167218e-03,  4.54646702e-13,
       -7.51897679e-13,  2.23088171e-11])
```

It seems like this issue is breaking attempts to use the ASE `SumCalculator` (which would conveniently allow someone to estimate the XC contribution similar to PBE accuracy using M3GNet and then just add the dispersion correction with DFT-D4).

**Comment by kenko911** [comment]
Hi @ajhoffman1229, sorry for the late reply. I tried the latest version of MatGL and it runs well without any errors. Please try to update the latest one. As for the second question, The difference between off-diagonal stress tensor is only about 1e-8 and many possible reasons can cause such small numerical noise involving multiple tensor operations, default precision and etc.....

**Comment by ajhoffman1229** [comment]
No problem, thanks for the quick reply. I'll update to the latest version and try again.

The second issue I encountered was an issue with the shape of the stress tensor, not the precision of its values, which is actually an issue with ASE. ASE's `get_property` function returns either the 3x3 stress tensor or a 6x1 array in Voigt notation. The defaults for different calculators differ, so there can be some bugs if using ASE's `SumCalculator` as a wrapper for M3GNet. I submitted a MR [here](https://gitlab.com/ase/ase/-/merge_requests/3028) to address the issue.
